File: python/algorithms/graph/shortest_path/dijkstra.py
"""Dijkstra 最短路。"""

from __future__ import annotations
import heapq


def dijkstra(adj: list[list[tuple[int, int]]], src: int) -> list[int]:
    """非负权图单源最短路；含负权边请用 Bellman–Ford。"""
    n = len(adj)
    for u in range(n):
        for _, w in adj[u]:
            if w < 0:
                raise ValueError("negative edge weight; use Bellman-Ford")
    dist = [10**18] * n
    dist[src] = 0
    pq: list[tuple[int, int]] = [(0, src)]
    while pq:
        d, u = heapq.heappop(pq)
        if d != dist[u]:
            continue
        for v, w in adj[u]:
            if dist[v] > d + w:
                dist[v] = d + w
                heapq.heappush(pq, (dist[v], v))
    return dist


if __name__ == "__main__":
    # 0--1 weight2, 0--2 weight5, 1--2 weight1
    g: list[list[tuple[int, int]]] = [[] for _ in range(3)]
    g[0].append((1, 2))
    g[1].append((0, 2))
    g[0].append((2, 5))
    g[2].append((0, 5))
    g[1].append((2, 1))
    g[2].append((1, 1))
    d = dijkstra(g, 0)
    assert d == [0, 2, 3]
    assert dijkstra([[]], 0) == [0]
    g2: list[list[tuple[int, int]]] = [[], [(0, 1)]]
    assert dijkstra(g2, 0)[1] == 10**18
    try:
        dijkstra([[(1, -1)], [(0, -1)]], 0)
        raise AssertionError("expected ValueError")
    except ValueError:
        pass
    print("shortest_path OK")

File: cpp/algorithms/graph/shortest_path/dijkstra.cpp
#include <bits/stdc++.h>
#include <cassert>
using namespace std;

vector<long long> dijkstra(const vector<vector<pair<int, int>>>& adj, int src) {
    int n = (int)adj.size();
    for (int u = 0; u < n; ++u)
        for (auto [v, w] : adj[u])
            if (w < 0) throw runtime_error("negative edge weight; use Bellman-Ford");
    const long long INF = (long long)4e18;
    vector<long long> dist(n, INF);
    dist[src] = 0;
    using P = pair<long long, int>;
    priority_queue<P, vector<P>, greater<P>> pq;
    pq.push({0, src});
    while (!pq.empty()) {
        auto [d, u] = pq.top();
        pq.pop();
        if (d != dist[u]) continue;
        for (auto [v, w] : adj[u]) {
            if (dist[v] > d + w) {
                dist[v] = d + w;
                pq.push({dist[v], v});
            }
        }
    }
    return dist;
}

int main() {
    vector<vector<pair<int, int>>> g(3);
    g[0].push_back({1, 2});
    g[1].push_back({0, 2});
    g[0].push_back({2, 5});
    g[2].push_back({0, 5});
    g[1].push_back({2, 1});
    g[2].push_back({1, 1});
    auto d = dijkstra(g, 0);
    assert(d == vector<long long>({0, 2, 3}));
    assert(dijkstra(vector<vector<pair<int, int>>>{vector<pair<int, int>>{}}, 0) == vector<long long>{0});
    vector<vector<pair<int, int>>> g2(2);
    g2[1].push_back({0, 1});
    assert(dijkstra(g2, 0)[1] == (long long)4e18);
    try {
        vector<vector<pair<int, int>>> neg(2);
        neg[0].push_back({1, -1});
        neg[1].push_back({0, -1});
        dijkstra(neg, 0);
        assert(false);
    } catch (const runtime_error&) {
    }
    cout << "shortest_path OK" << endl;
    return 0;
}
